`implied_bounds_in_impls` false positive

### Summary

The lint suggests removing a specified bound, however, removing the bound breaks code that then uses the function.

### Lint Name

implied_bounds_in_impls

### Reproducer

I tried this code:

```rust
use std::{
    cell::RefCell,
    fmt,
    ops::{Deref, DerefMut},
};

pub struct HeavyState<T: fmt::Debug> {
    current: RefCell<T>,
}

impl<T: fmt::Debug> HeavyState<T> {
    pub fn get_mut(&self) -> impl Deref<Target = T> + DerefMut + '_ {
        self.current.borrow_mut()
    }
}

pub fn foo(bindings: HeavyState<Vec<u32>>) {
   let mut b = bindings.get_mut();
   b.push(0);
}

fn main() {}
```

Clippy reports the following:

```
warning: this bound is already specified as the supertrait of `DerefMut`
  --> src/main.rs:14:35
   |
14 |     pub fn get_mut(&self) -> impl Deref<Target = T> + DerefMut + '_ {
   |                                   ^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implied_bounds_in_impls
   = note: `#[warn(clippy::implied_bounds_in_impls)]` on by default
help: try removing this bound
   |
14 -     pub fn get_mut(&self) -> impl Deref<Target = T> + DerefMut + '_ {
14 +     pub fn get_mut(&self) -> impl DerefMut + '_ {
   |
```

However, applying the suggestion results in the following:
```
error[E0599]: no method named `push` found for opaque type `impl std::ops::DerefMut + '_` in the current scope
  --> src/main.rs:23:7
   |
23 |     b.push(0);
   |       ^^^^ method not found in `impl DerefMut`
```

The suggestion should be to transfer the associated type bound `Target = T` onto `DerefMut` rather than entirely removing the bound. This would be the correct code:
```rs
impl<T: fmt::Debug> HeavyState<T> {
    pub fn get_mut(&self) -> impl DerefMut<Target = T> + '_ {
        self.current.borrow_mut()
    }
}
```

### Version

```text
rustc 1.74.0-nightly (84a9f4c6e 2023-08-29)
binary: rustc
commit-hash: 84a9f4c6e6f0f02bff9acc9a5e0305a1554a87fc
commit-date: 2023-08-29
host: x86_64-apple-darwin
release: 1.74.0-nightly
LLVM version: 17.0.0
```


### Additional Labels

@rustbot label +I-suggestion-causes-error

--- a/clippy_lints/src/implied_bounds_in_impls.rs
+++ b/clippy_lints/src/implied_bounds_in_impls.rs
@@ -56,7 +56,9 @@ fn emit_lint(
     poly_trait: &rustc_hir::PolyTraitRef<'_>,
     opaque_ty: &rustc_hir::OpaqueTy<'_>,
     index: usize,
+    // The bindings that were implied
     implied_bindings: &[rustc_hir::TypeBinding<'_>],
+    // The original bindings that `implied_bindings` are implied from
     implied_by_bindings: &[rustc_hir::TypeBinding<'_>],
     implied_by_args: &[GenericArg<'_>],
     implied_by_span: Span,
@@ -69,7 +71,7 @@ fn emit_lint(
         poly_trait.span,
         &format!("this bound is already specified as the supertrait of `{implied_by}`"),
         |diag| {
-            // If we suggest removing a bound, we may also need extend the span
+            // If we suggest removing a bound, we may also need to extend the span
             // to include the `+` token that is ahead or behind,
             // so we don't end up with something like `impl + B` or `impl A + `
 

--- a/clippy_lints/src/implied_bounds_in_impls.rs
+++ b/clippy_lints/src/implied_bounds_in_impls.rs
@@ -50,6 +50,93 @@ declare_clippy_lint! {
 }
 declare_lint_pass!(ImpliedBoundsInImpls => [IMPLIED_BOUNDS_IN_IMPLS]);
 
+#[allow(clippy::too_many_arguments)]
+fn emit_lint(
+    cx: &LateContext<'_>,
+    poly_trait: &rustc_hir::PolyTraitRef<'_>,
+    opaque_ty: &rustc_hir::OpaqueTy<'_>,
+    index: usize,
+    implied_bindings: &[rustc_hir::TypeBinding<'_>],
+    implied_by_bindings: &[rustc_hir::TypeBinding<'_>],
+    implied_by_args: &[GenericArg<'_>],
+    implied_by_span: Span,
+) {
+    let implied_by = snippet(cx, implied_by_span, "..");
+
+    span_lint_and_then(
+        cx,
+        IMPLIED_BOUNDS_IN_IMPLS,
+        poly_trait.span,
+        &format!("this bound is already specified as the supertrait of `{implied_by}`"),
+        |diag| {
+            // If we suggest removing a bound, we may also need extend the span
+            // to include the `+` token that is ahead or behind,
+            // so we don't end up with something like `impl + B` or `impl A + `
+
+            let implied_span_extended = if let Some(next_bound) = opaque_ty.bounds.get(index + 1) {
+                poly_trait.span.to(next_bound.span().shrink_to_lo())
+            } else if index > 0
+                && let Some(prev_bound) = opaque_ty.bounds.get(index - 1)
+            {
+                prev_bound.span().shrink_to_hi().to(poly_trait.span.shrink_to_hi())
+            } else {
+                poly_trait.span
+            };
+
+            let mut sugg = vec![(implied_span_extended, String::new())];
+
+            // We also might need to include associated type binding that were specified in the implied bound,
+            // but omitted in the implied-by bound:
+            // `fn f() -> impl Deref<Target = u8> + DerefMut`
+            // If we're going to suggest removing `Deref<..>`, we'll need to put `<Target = u8>` on `DerefMut`
+            let omitted_assoc_tys: Vec<_> = implied_bindings
+                .iter()
+                .filter(|binding| !implied_by_bindings.iter().any(|b| b.ident == binding.ident))
+                .collect();
+
+            if !omitted_assoc_tys.is_empty() {
+                // `<>` needs to be added if there aren't yet any generic arguments or bindings
+                let needs_angle_brackets = implied_by_args.is_empty() && implied_by_bindings.is_empty();
+                let insert_span = match (implied_by_args, implied_by_bindings) {
+                    ([.., arg], [.., binding]) => arg.span().max(binding.span).shrink_to_hi(),
+                    ([.., arg], []) => arg.span().shrink_to_hi(),
+                    ([], [.., binding]) => binding.span.shrink_to_hi(),
+                    ([], []) => implied_by_span.shrink_to_hi(),
+                };
+
+                let mut associated_tys_sugg = if needs_angle_brackets {
+                    "<".to_owned()
+                } else {
+                    // If angle brackets aren't needed (i.e., there are already generic arguments or bindings),
+                    // we need to add a comma:
+                    // `impl A<B, C >`
+                    //             ^ if we insert `Assoc=i32` without a comma here, that'd be invalid syntax:
+                    // `impl A<B, C Assoc=i32>`
+                    ", ".to_owned()
+                };
+
+                for (index, binding) in omitted_assoc_tys.into_iter().enumerate() {
+                    if index > 0 {
+                        associated_tys_sugg += ", ";
+                    }
+                    associated_tys_sugg += &snippet(cx, binding.span, "..");
+                }
+                if needs_angle_brackets {
+                    associated_tys_sugg += ">";
+                }
+                sugg.push((insert_span, associated_tys_sugg));
+            }
+
+            diag.multipart_suggestion_with_style(
+                "try removing this bound",
+                sugg,
+                Applicability::MachineApplicable,
+                SuggestionStyle::ShowAlways,
+            );
+        },
+    );
+}
+
 /// Tries to "resolve" a type.
 /// The index passed to this function must start with `Self=0`, i.e. it must be a valid
 /// type parameter index.
@@ -189,80 +276,15 @@ fn check(cx: &LateContext<'_>, decl: &FnDecl<'_>) {
                         })
                     })
             {
-                let implied_by = snippet(cx, implied_by_span, "..");
-                span_lint_and_then(
-                    cx, IMPLIED_BOUNDS_IN_IMPLS,
-                    poly_trait.span,
-                    &format!("this bound is already specified as the supertrait of `{implied_by}`"),
-                    |diag| {
-                        // If we suggest removing a bound, we may also need extend the span
-                        // to include the `+` token that is ahead or behind,
-                        // so we don't end up with something like `impl + B` or `impl A + `
-
-                        let implied_span_extended = if let Some(next_bound) = opaque_ty.bounds.get(index + 1) {
-                            poly_trait.span.to(next_bound.span().shrink_to_lo())
-                        } else if index > 0
-                            && let Some(prev_bound) = opaque_ty.bounds.get(index - 1)
-                        {
-                            prev_bound.span().shrink_to_hi().to(poly_trait.span.shrink_to_hi())
-                        } else {
-                            poly_trait.span
-                        };
-
-                        let mut sugg = vec![
-                            (implied_span_extended, String::new()),
-                        ];
-
-                        // We also might need to include associated type binding that were specified in the implied bound,
-                        // but omitted in the implied-by bound:
-                        // `fn f() -> impl Deref<Target = u8> + DerefMut`
-                        // If we're going to suggest removing `Deref<..>`, we'll need to put `<Target = u8>` on `DerefMut`
-                        let omitted_assoc_tys: Vec<_> = implied_bindings
-                            .iter()
-                            .filter(|binding| {
-                                implied_by_bindings
-                                    .iter()
-                                    // TODO: is checking idents enough for stuff like `<Target: Sized> == <Target = u8>`
-                                    .find(|b| b.ident == binding.ident)
-                                    .is_none()
-                            })
-                            .collect();
-                        
-                        if !omitted_assoc_tys.is_empty() {
-                            // `<>` needs to be added if there aren't yet any generic arguments or bindings
-                            let needs_angle_brackets = implied_by_args.is_empty() && implied_by_bindings.is_empty();
-                            let insert_span = match (implied_by_args, implied_by_bindings) {
-                                ([.., arg], [.., binding]) => arg.span().max(binding.span).shrink_to_hi(),
-                                ([.., arg], []) => arg.span().shrink_to_hi(),
-                                ([], [.., binding]) => binding.span.shrink_to_hi(),
-                                ([], []) => implied_by_span.shrink_to_hi(),
-                            };
-
-                            let mut associated_tys_sugg = if needs_angle_brackets {
-                                "<".to_owned()
-                            } else {
-                                // If angle brackets aren't needed (i.e., there are already generic arguments or bindings),
-                                // we need to add a comma:
-                                // `impl A<B, C >`
-                                //             ^ if we insert `Assoc=i32` without a comma here, that'd be invalid syntax:
-                                // `impl A<B, C Assoc=i32>`
-                                ", ".to_owned()
-                            };
-
-                            for (index, binding) in omitted_assoc_tys.into_iter().enumerate() {
-                                if index > 0 {
-                                    associated_tys_sugg += ", ";
-                                }
-                                associated_tys_sugg += &snippet(cx, binding.span, "..");
-                            }
-                            if needs_angle_brackets {
-                                associated_tys_sugg += ">";
-                            }
-                            sugg.push((insert_span, associated_tys_sugg));
-                        }
-
-                        diag.multipart_suggestion_with_style("try removing this bound", sugg, Applicability::MachineApplicable, SuggestionStyle::ShowAlways);
-                    }
+                emit_lint(
+                    cx,
+                    poly_trait,
+                    opaque_ty,
+                    index,
+                    implied_bindings,
+                    implied_by_bindings,
+                    implied_by_args,
+                    implied_by_span
                 );
             }
         }

--- a/clippy_lints/src/implied_bounds_in_impls.rs
+++ b/clippy_lints/src/implied_bounds_in_impls.rs
@@ -149,7 +149,7 @@ fn check(cx: &LateContext<'_>, decl: &FnDecl<'_>) {
                 && let predicates = cx.tcx.super_predicates_of(trait_def_id).predicates
                 && !predicates.is_empty() // If the trait has no supertrait, there is nothing to add.
             {
-                Some((bound.span(), path.args.map_or([].as_slice(), |a| a.args), predicates, trait_def_id))
+                Some((bound.span(), path, predicates, trait_def_id))
             } else {
                 None
             }
@@ -162,10 +162,14 @@ fn check(cx: &LateContext<'_>, decl: &FnDecl<'_>) {
             if let GenericBound::Trait(poly_trait, TraitBoundModifier::None) = bound
                 && let [.., path] = poly_trait.trait_ref.path.segments
                 && let implied_args = path.args.map_or([].as_slice(), |a| a.args)
+                && let implied_bindings = path.args.map_or([].as_slice(), |a| a.bindings)
                 && let Some(def_id) = poly_trait.trait_ref.path.res.opt_def_id()
-                && let Some(implied_by_span) = implied_bounds
+                && let Some((implied_by_span, implied_by_args, implied_by_bindings)) = implied_bounds
                     .iter()
-                    .find_map(|&(span, implied_by_args, preds, implied_by_def_id)| {
+                    .find_map(|&(span, implied_by_path, preds, implied_by_def_id)| {
+                        let implied_by_args = implied_by_path.args.map_or([].as_slice(), |a| a.args);
+                        let implied_by_bindings = implied_by_path.args.map_or([].as_slice(), |a| a.bindings);
+
                         preds.iter().find_map(|(clause, _)| {
                             if let ClauseKind::Trait(tr) = clause.kind().skip_binder()
                                 && tr.def_id() == def_id
@@ -178,7 +182,7 @@ fn check(cx: &LateContext<'_>, decl: &FnDecl<'_>) {
                                     def_id,
                                 )
                             {
-                                Some(span)
+                                Some((span, implied_by_args, implied_by_bindings))
                             } else {
                                 None
                             }
@@ -192,21 +196,72 @@ fn check(cx: &LateContext<'_>, decl: &FnDecl<'_>) {
                     &format!("this bound is already specified as the supertrait of `{implied_by}`"),
                     |diag| {
                         // If we suggest removing a bound, we may also need extend the span
-                        // to include the `+` token, so we don't end up with something like `impl + B`
+                        // to include the `+` token that is ahead or behind,
+                        // so we don't end up with something like `impl + B` or `impl A + `
 
                         let implied_span_extended = if let Some(next_bound) = opaque_ty.bounds.get(index + 1) {
                             poly_trait.span.to(next_bound.span().shrink_to_lo())
+                        } else if index > 0
+                            && let Some(prev_bound) = opaque_ty.bounds.get(index - 1)
+                        {
+                            prev_bound.span().shrink_to_hi().to(poly_trait.span.shrink_to_hi())
                         } else {
                             poly_trait.span
                         };
 
-                        diag.span_suggestion_with_style(
-                            implied_span_extended,
-                            "try removing this bound",
-                            "",
-                            Applicability::MachineApplicable,
-                            SuggestionStyle::ShowAlways
-                        );
+                        let mut sugg = vec![
+                            (implied_span_extended, String::new()),
+                        ];
+
+                        // We also might need to include associated type binding that were specified in the implied bound,
+                        // but omitted in the implied-by bound:
+                        // `fn f() -> impl Deref<Target = u8> + DerefMut`
+                        // If we're going to suggest removing `Deref<..>`, we'll need to put `<Target = u8>` on `DerefMut`
+                        let omitted_assoc_tys: Vec<_> = implied_bindings
+                            .iter()
+                            .filter(|binding| {
+                                implied_by_bindings
+                                    .iter()
+                                    // TODO: is checking idents enough for stuff like `<Target: Sized> == <Target = u8>`
+                                    .find(|b| b.ident == binding.ident)
+                                    .is_none()
+                            })
+                            .collect();
+                        
+                        if !omitted_assoc_tys.is_empty() {
+                            // `<>` needs to be added if there aren't yet any generic arguments or bindings
+                            let needs_angle_brackets = implied_by_args.is_empty() && implied_by_bindings.is_empty();
+                            let insert_span = match (implied_by_args, implied_by_bindings) {
+                                ([.., arg], [.., binding]) => arg.span().max(binding.span).shrink_to_hi(),
+                                ([.., arg], []) => arg.span().shrink_to_hi(),
+                                ([], [.., binding]) => binding.span.shrink_to_hi(),
+                                ([], []) => implied_by_span.shrink_to_hi(),
+                            };
+
+                            let mut associated_tys_sugg = if needs_angle_brackets {
+                                "<".to_owned()
+                            } else {
+                                // If angle brackets aren't needed (i.e., there are already generic arguments or bindings),
+                                // we need to add a comma:
+                                // `impl A<B, C >`
+                                //             ^ if we insert `Assoc=i32` without a comma here, that'd be invalid syntax:
+                                // `impl A<B, C Assoc=i32>`
+                                ", ".to_owned()
+                            };
+
+                            for (index, binding) in omitted_assoc_tys.into_iter().enumerate() {
+                                if index > 0 {
+                                    associated_tys_sugg += ", ";
+                                }
+                                associated_tys_sugg += &snippet(cx, binding.span, "..");
+                            }
+                            if needs_angle_brackets {
+                                associated_tys_sugg += ">";
+                            }
+                            sugg.push((insert_span, associated_tys_sugg));
+                        }
+
+                        diag.multipart_suggestion_with_style("try removing this bound", sugg, Applicability::MachineApplicable, SuggestionStyle::ShowAlways);
                     }
                 );
             }

--- a/tests/ui/implied_bounds_in_impls.fixed
+++ b/tests/ui/implied_bounds_in_impls.fixed
@@ -83,4 +83,43 @@ mod issue11422 {
     fn f() -> impl Trait1<()> + Trait2 {}
 }
 
+mod issue11435 {
+    // Associated type needs to be included on DoubleEndedIterator in the suggestion
+    fn my_iter() -> impl DoubleEndedIterator<Item = u32> {
+        0..5
+    }
+
+    // Removing the `Clone` bound should include the `+` behind it in its remove suggestion
+    fn f() -> impl Copy {
+        1
+    }
+
+    trait Trait1<T> {
+        type U;
+    }
+    impl Trait1<i32> for () {
+        type U = i64;
+    }
+    trait Trait2<T>: Trait1<T> {}
+    impl Trait2<i32> for () {}
+
+    // When the other trait has generics, it shouldn't add another pair of `<>`
+    fn f2() -> impl Trait2<i32, U = i64> {}
+
+    trait Trait3<T, U, V> {
+        type X;
+        type Y;
+    }
+    trait Trait4<T>: Trait3<T, i16, i64> {}
+    impl Trait3<i8, i16, i64> for () {
+        type X = i32;
+        type Y = i128;
+    }
+    impl Trait4<i8> for () {}
+
+    // Associated type `X` is specified, but `Y` is not, so only that associated type should be moved
+    // over
+    fn f3() -> impl Trait4<i8, X = i32, Y = i128> {}
+}
+
 fn main() {}

--- a/tests/ui/implied_bounds_in_impls.rs
+++ b/tests/ui/implied_bounds_in_impls.rs
@@ -83,4 +83,43 @@ mod issue11422 {
     fn f() -> impl Trait1<()> + Trait2 {}
 }
 
+mod issue11435 {
+    // Associated type needs to be included on DoubleEndedIterator in the suggestion
+    fn my_iter() -> impl Iterator<Item = u32> + DoubleEndedIterator {
+        0..5
+    }
+
+    // Removing the `Clone` bound should include the `+` behind it in its remove suggestion
+    fn f() -> impl Copy + Clone {
+        1
+    }
+
+    trait Trait1<T> {
+        type U;
+    }
+    impl Trait1<i32> for () {
+        type U = i64;
+    }
+    trait Trait2<T>: Trait1<T> {}
+    impl Trait2<i32> for () {}
+
+    // When the other trait has generics, it shouldn't add another pair of `<>`
+    fn f2() -> impl Trait1<i32, U = i64> + Trait2<i32> {}
+
+    trait Trait3<T, U, V> {
+        type X;
+        type Y;
+    }
+    trait Trait4<T>: Trait3<T, i16, i64> {}
+    impl Trait3<i8, i16, i64> for () {
+        type X = i32;
+        type Y = i128;
+    }
+    impl Trait4<i8> for () {}
+
+    // Associated type `X` is specified, but `Y` is not, so only that associated type should be moved
+    // over
+    fn f3() -> impl Trait3<i8, i16, i64, X = i32, Y = i128> + Trait4<i8, X = i32> {}
+}
+
 fn main() {}

--- a/tests/ui/implied_bounds_in_impls.stderr
+++ b/tests/ui/implied_bounds_in_impls.stderr
@@ -143,5 +143,53 @@ LL -     fn default_generic_param2() -> impl PartialOrd + PartialEq + Debug {}
 LL +     fn default_generic_param2() -> impl PartialOrd + Debug {}
    |
 
-error: aborting due to 12 previous errors
+error: this bound is already specified as the supertrait of `DoubleEndedIterator`
+  --> $DIR/implied_bounds_in_impls.rs:88:26
+   |
+LL |     fn my_iter() -> impl Iterator<Item = u32> + DoubleEndedIterator {
+   |                          ^^^^^^^^^^^^^^^^^^^^
+   |
+help: try removing this bound
+   |
+LL -     fn my_iter() -> impl Iterator<Item = u32> + DoubleEndedIterator {
+LL +     fn my_iter() -> impl DoubleEndedIterator<Item = u32> {
+   |
+
+error: this bound is already specified as the supertrait of `Copy`
+  --> $DIR/implied_bounds_in_impls.rs:93:27
+   |
+LL |     fn f() -> impl Copy + Clone {
+   |                           ^^^^^
+   |
+help: try removing this bound
+   |
+LL -     fn f() -> impl Copy + Clone {
+LL +     fn f() -> impl Copy {
+   |
+
+error: this bound is already specified as the supertrait of `Trait2<i32>`
+  --> $DIR/implied_bounds_in_impls.rs:107:21
+   |
+LL |     fn f2() -> impl Trait1<i32, U = i64> + Trait2<i32> {}
+   |                     ^^^^^^^^^^^^^^^^^^^^
+   |
+help: try removing this bound
+   |
+LL -     fn f2() -> impl Trait1<i32, U = i64> + Trait2<i32> {}
+LL +     fn f2() -> impl Trait2<i32, U = i64> {}
+   |
+
+error: this bound is already specified as the supertrait of `Trait4<i8, X = i32>`
+  --> $DIR/implied_bounds_in_impls.rs:122:21
+   |
+LL |     fn f3() -> impl Trait3<i8, i16, i64, X = i32, Y = i128> + Trait4<i8, X = i32> {}
+   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+help: try removing this bound
+   |
+LL -     fn f3() -> impl Trait3<i8, i16, i64, X = i32, Y = i128> + Trait4<i8, X = i32> {}
+LL +     fn f3() -> impl Trait4<i8, X = i32, Y = i128> {}
+   |
+
+error: aborting due to 16 previous errors
 

